patches/minecraft/net/minecraft/client/network/play/NetworkPlayerInfo.java.patch

```diff
@@ -0,0 +1,12 @@
+--- a/net/minecraft/client/network/play/NetworkPlayerInfo.java
++++ b/net/minecraft/client/network/play/NetworkPlayerInfo.java
+@@ -50,7 +50,8 @@
+    }
+ 
+    protected void func_178839_a(GameType p_178839_1_) {
+-      this.field_178866_b = p_178839_1_;
++       if(Minecraft.func_71410_x().field_71439_g != null && this.field_178867_a.getId().equals(Minecraft.func_71410_x().field_71439_g.func_110124_au()) && this.field_178866_b != p_178839_1_) net.minecraftforge.common.ForgeHooks.onClientChangeGameMode(Minecraft.func_71410_x().field_71439_g, this.field_178866_b, p_178839_1_);
++       this.field_178866_b = p_178839_1_;
```

Comment:
Indentation still wrong. Also missing space after `if`.
Also, my point of separating client and server event was so that you can fire the event on the client _always_, by not including the `PlayerEntity`.

Reply:
When the event is fired client side without the player check there are two problems:

1. The event is fired for all other tracked players of the local client. If a player is too far away it won't be fired.
2. Out of the tracked players only the ones not in spectator are fired. If another player comes out of spectator mode the local client won't know about the change and the event is not fired.
3. Sometimes neither of these conditions are met and the event *still* fires.
4. A reliable alternative is to just have the mod send a packet to the client about their mod specific status updates. Because the server knows about all players, the mod dev can control the scope of the event.

By adding the player, the goal is to add consistent results for when the event is fired.

Reply:
That is not true. The packet is sent to all players.
Have you actually verified these claims you are making?

Reply:
I just redid me experiment and found different results. I think it may have been due to poor placement of one of the patches but my claims are not true. All players are always notified of all game mode changes, even for players they have never or currently cannot see.